[v1.10] Fix race condition in DNS proxy when multiple DNS requests for the same name end up with policy drops

michi-covalent opened the pull request:
* #22252 -- daemon, fqdn: Fix race between parallel DNS requests for the same name (@christarazi)

Once this PR is merged, you can update the PR labels via:
```upstream-prs
$ for pr in 22252; do contrib/backporting/set-labels.py $pr done 1.10; done
```

christarazi committed:
daemon, fqdn: Fix race between parallel DNS requests for the same name

[ upstream commit TODO ]

It is possible for parallel DNS requests handled by separate goroutines
(controlled by --dnsproxy-concurrency-limit) to race if they are for the
same DNS name, i.e. both goroutines handling cilium.io.

Here is a diagram for this race:

```
             G1                                    G2

T0 --> NotifyOnDNSMsg()               NotifyOnDNSMsg()            <-- T0

T1 --> UpdateGenerateDNS()            UpdateGenerateDNS()         <-- T1

T2 ----> mutex.Lock()                 +---------------------------+
                                      |No identities need updating|
T3 ----> mutex.Unlock()               +---------------------------+

T4 --> UpsertGeneratedIdentities()    UpsertGeneratedIdentities() <-- T4

T5 ---->  Upsert()                    DNS released back to pod    <-- T5
                                                   |
T6 --> DNS released back to pod                    |
             |                                     |
             |                                     |
             v                                     v
      Traffic flows fine                   Leads to policy drop
```

Note how G2 releases the DNS msg back to the pod at T5 because
UpdateGenerateDNS() was a no-op. It's a no-op because G1 had executed
UpdateGenerateDNS() first at T1 and performed the necessary identity
allocation for the response IPs. Due to G1 performing all the work
first, G2 executes T4 also as a no-op and releases the msg back to the
pod at T5 before G1 would at T6.

The above results in a policy drop because it is assumed that when the
DNS msg is released back to the pod / endpoint, then the ipcache has
been updated with the proper IP <-> identity mapping. In other words,
because G2 releases the DNS msg back to the pod / endpoint, the app
attempts to use the IPs returned in the DNS response, but the ipcache
lookup fails in the datapath because the IP <-> identity mappings don't
exist. In essence, this is a race between DNS requests to the ipcache.

To fix, we create a critical section when handling the DNS response IPs.
Allocating identities and associating them to IPs in the ipcache is now
atomic because of the critical section. This is especially important as
we've already seen for multiple DNS requests are in-flight for the same
name (i.e. cilium.io).

The critical section is implemented using a fixed sized array of
mutexes. Why? Because the other options (listed below) all have flaws
which may result in drops.

Other options considered:

1) A map of mutexes keyed on DNS name
2) A map of mutexes keyed on set of DNS response IPs
3) A map of mutexes keyed on endpoint ID
4) A slice of mutexes with hash based on individual IPs from the set of
   DNS response IPs

(1) could cause a race between the DNS response IPs from `a.com` and the
set of DNS response IPs from `b.com` especially if the sets are the
same.
(2) could race between a super / subset of DNS response IPs, i.e. only
an intersection of sets would have protection. Any IPs that are not in
the intersection are vulnerable to race.
(3) could cause a race between different endpoints querying the same
name, i.e. two endpoints querying cilium.io.
(4) could work but this is approach is not ideal as the slice scales
with the number of unique IPs seen by the DNS proxy. This has memory
consumption concerns.

Therefore, we opt for a fixed size array of mutexes where we hash the
individual IPs from the set of DNS response IPs. For example, say we
attempt to resolve cilium.io and the response IPs are A and B.

  hash(A) % mod = 8
  hash(B) % mod = 3

  where mod is --dnsproxy-lock-size, i.e. size of array

What's returned is a slice containing the above array indices that's
sorted in ascending order. This is the order in which the mutexes must
be locked and unlocked. To be specific, [3, 8] is the sorted resulted
and it means we acquire & release mutex at index 3 and then do the same
for mutex at index 8.

What this does is essentially serialize parallel DNS requests which
involve the same IP <-> identity mappings in the ipcache.

It's possible that the hash of the IPs collide (map to the same mutex)
when the IPs are different, but this is OK because it's overprotecting,
rather then underprotecting which the other options above suffer from.
Overprotecting results in a small performance hit because we are
potentially serializing completely different / unrelated DNS requests,
i.e. no intersection in the response IP sets.

For --dnsproxy-lock-size, a default of 128 was chosen as a reasonable
balance between memory usage and hash collisions, for most deployments
of Cilium. An increased lock size will result in more memory usage, but
faster DNS processing as there are less IP hash collisions, and
therefore less mutexes to acquire. A decreased lock size will save
memory at the cost of slower DNS processing as the IP hash collisions
are more likely, and this tends towards behavior similar to a single,
shared mutex. Users who need to tune this value because they have many
parallel DNS requests (1000s) can make the trade-off of using more
memory to avoid hash collisions. It's worth mentioning that the memory
usage is for storing a instance of a mutex which is 64 bits as of Go
1.19 [1].

Benchmark run comparing the code before this commit, changing the fixed
sized array to one to simulate a single mutex, and this commit
(multiple-locks):

```
$ go test -v -tags integration_tests ./daemon/cmd -check.b -check.bmem -check.f DaemonFQDNSuite.Benchmark_notifyOnDNSMsg -test.benchtime 100x -test.count 8
```

```
$ benchstat no-lock.txt single-lock.txt multiple-locks.txt
name \ time/op    no-lock.txt  single-lock.txt  multiple-locks.txt
_notifyOnDNSMsg   93.2µs ± 9%     179.6µs ± 3%        143.2µs ±28%

name \ alloc/op   no-lock.txt  single-lock.txt  multiple-locks.txt
_notifyOnDNSMsg   36.5kB ± 2%      35.3kB ± 0%         33.8kB ± 1%

name \ allocs/op  no-lock.txt  single-lock.txt  multiple-locks.txt
_notifyOnDNSMsg      485 ± 1%         474 ± 0%            453 ± 1%
```

Reproducing the bug and performance testing on a real cluster included
running the 3 different modes and comparing the average processing time
of the DNS proxy code. The cluster was 2 nodes with 32 replicas of the
following workload:

  # websites.txt contains different common URLs, some duplicated.
  while [[ true ]]; do
    curl --parallel --parallel-immediate --parallel-max 5 --config websites.txt
    sleep 0.25s
  done

No lock:        4.61ms, 18.3ms
Single lock:    7.42ms, 18.3ms
Multiple locks: 5.25ms, 17.9ms

[1]: https://cs.opensource.google/go/go/+/refs/tags/go1.19.3:src/sync/mutex.go;l=34

Co-authored-by: Michi Mutsuzaki <michi@isovalent.com>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

joestringer commented on daemon/cmd/fqdn.go:
```diff
@@ -576,6 +576,39 @@ func (d *Daemon) notifyOnDNSMsg(lookupTime time.Time, ep *endpoint.Endpoint, epI
+		// G1 performing all the work first, G2 executes T4 as a no-op and
+		// releases the msg back to the app at T5 before G1 can.
+		d.notifyOnDNSMsgMu.Lock()
+		defer d.notifyOnDNSMsgMu.Unlock()
```
This is serializing *all* DNS requests, is that necessary to fix the issue for DNS requests for the same name? Do you think that it is likely that this request handling is already serialized somewhat by the code patterns or will this be a performance regression? (For background, we use this fancy framework that splits DNS handling out into a bunch of independent goroutines so it seems a bit odd to then force the expensive part of handling the DNS responses to now be 100% serialized)

michi-covalent replied:
no it's not necessary, i think we just did the simplest thing to see if it would fix the issue. we also talked about a fixed size list of mutexes and hash dns names to map to mutexes or something like that.

chris is doing perf testing right now 🧑‍🔬

christarazi replied:
Fixed size list of mutexes implemented in last commit.

christarazi committed:
daemon, cmd: Add benchmark for notifyOnDNSMsg()

[ upstream commit TODO]

[ Backporter's notes: This commit contains slightly modified code to the
                      upstream version, as it introduces
                      proxy.SetEndpointLookuper in order to facilitate
                      running the benchmark added in this commit. In the
                      upstream version, the log recorder no longer does
                      an endpoint lookup in notifyOnDNSMsg(), so we
                      don't need to set proxy.endpointManager. ]

This allows us to stress / benchmark the main callback function of the
DNS proxy within Cilium. This callback is called on each DNS request and
response made by each endpoint managed by Cilium.

For this commit, we are only considering the DNS response path because
it is the path which contains the most work to be done. This includes
policy calculation, identity allocation, and ipcache upsert.

It is useful to establish baseline performance numbers so that we can
compare against changes to this code path. In the following commits, a
mutex will be added to this path to fix a race condition during parallel
DNS response handling.

Baseline:

```
$ go test -v -tags integration_tests ./daemon/cmd -check.b -check.bmem -check.f DaemonFQDNSuite.Benchmark_notifyOnDNSMsg -test.benchtime 100x
...
PASS: fqdn_test.go:180: DaemonFQDNSuite.Benchmark_notifyOnDNSMsg           20000             96078 ns/op        36567 B/op         482 allocs/op
```

Signed-off-by: Chris Tarazi <chris@isovalent.com>
